Core/ModuleInstaller.cs

```diff
@@ -1100,6 +1100,90 @@ public void Upgrade(IEnumerable<CkanModule> modules, NetAsyncModulesDownloader n
+                    //Maybe ModuleNotInstalled ?
+                    if (registry_manager.registry.IsAutodetected(ident))
+                    {
+                        throw new ModuleNotFoundKraken(ident, repl.ToReplace.version.ToString(), String.Format("Can't replace {0} as it was not installed by CKAN. \r\n Please remove manually before trying to install it.", ident));
+                    }
+
+                    throw new ModuleNotFoundKraken(ident, repl.ToReplace.version.ToString(), String.Format("Can't replace {0} as it is not installed. Please attempt to install {1} instead.", ident, repl.ReplaceWith.identifier));
+                }
+                else
+                {
+                    // Obviously, we need to remove the mod we are replacing
+                    to_remove.Add(repl.ToReplace.identifier);
+
+                    log.DebugFormat("Ok, we are removing {0}", repl.ToReplace.identifier);
+                    //Check whether our Replacement target is already installed
+                    InstalledModule installed_replacement = registry_manager.registry.InstalledModule(repl.ReplaceWith.identifier);
+
+                    // If replacement is not installed, we've already added it to modsToInstall above
+                    if (installed_replacement != null)
+                    {
+                        //Module already installed. We'll need to treat it as an upgrade.
+                        log.DebugFormat("It turns out {0} is already installed, we'll upgrade it.", installed_replacement.identifier);
+                        to_remove.Add(installed_replacement.identifier);
+
+                        CkanModule installed = installed_replacement.Module;
+                        if (installed.version.IsEqualTo(repl.ReplaceWith.version))
+                        {
+                            log.InfoFormat("{0} is already at the latest version, reinstalling to replace {1}", repl.ReplaceWith.identifier, repl.ToReplace.identifier);
+                        }
+                        else if (installed.version.IsGreaterThan(repl.ReplaceWith.version))
+                        {
+                            log.WarnFormat("Downgrading {0} from {1} to {2} to replace {3}", repl.ReplaceWith.identifier, repl.ReplaceWith.version, repl.ReplaceWith.version, repl.ToReplace.identifier);
+                        }
+                        else
+                        {
+                            log.InfoFormat("Upgrading {0} to {1} to replace {2}", repl.ReplaceWith.identifier, repl.ReplaceWith.version, repl.ToReplace.identifier);
+                        }
+                    }
+                    else
+                    {
+                        log.InfoFormat("Replacing {0} with {1} {2}", repl.ToReplace.identifier, repl.ReplaceWith.identifier, repl.ReplaceWith.version);
+                    }
+                }
+            }
+
+            AddRemove(
+                modsToInstall,
+                to_remove,
```

Comment:
Minor style thing, but it would be nice to name these consistently. I.e., either:
```
modsToInstall
modsToRemove
```
or:
```
to_install
to_remove
```